web_ui/prometheus.go

```diff
@@ -138,6 +143,83 @@ func runtimeInfo() (api_v1.RuntimeInfo, error) {
 	return api_v1.RuntimeInfo{}, nil
 }
 
+func checkPromToken(av1 *route.Router) gin.HandlerFunc {
+	/* A function which wraps around the av1 router to force a jwk token check using
+	 * the origin's private key. It will check the request's URL and Header for a token
+	 * and if found it will then attempt to validate the token. If valid, it will continue
+	 * the routing as normal, otherwise it will return an error"
+	 */
+	return func(c *gin.Context) {
+		req := c.Request
+
+		var strToken string
+		var token jwt.Token
+		var err error
+		if authzQuery := req.URL.Query()["authz"]; len(authzQuery) > 0 {
+			strToken = authzQuery[0]
+		} else if authzHeader := req.Header["Authorization"]; len(authzHeader) > 0 {
+			strToken = strings.TrimPrefix(authzHeader[0], "Bearer ")
+		} else {
+			c.JSON(403, gin.H{"error": "Permission Denied: Missing token"})
+			return
+		}
+
+		token, err = jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+
+		if err != nil {
+			c.JSON(400, gin.H{"error": "Failed to parse token"})
+		}
+
+		fedURL := viper.GetString("FederationURL")
+
+		var bKey *jwk.Key
+		if fedURL == token.Issuer() {
+			err := pelican_config.DiscoverFederation()
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to discover the federation information"})
+			}
+			fedURIFile := viper.GetString("FederationURI")
+			response, err := http.Get(fedURIFile)
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to get federation key file"})
+			}
+			defer response.Body.Close()
+			contents, err := io.ReadAll(response.Body)
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to read federation key file"})
+				return
+			}
+			key, err := jwk.ParseKey(contents, jwk.WithPEM(true))
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to parse Federation key file"})
+				return
+			}
+			bKey = &key
+		} else {
+			bKey, err = pelican_config.GetOriginJWK()
+			if err != nil {
+				c.JSON(400, gin.H{"error": "Failed to retrieve private key"})
+				return
+			}
+		}
+
+		var raw ecdsa.PrivateKey
+		if err = (*bKey).Raw(&raw); err != nil {
+			c.JSON(400, gin.H{"error": "Failed to extract signing key"})
+			return
+		}
+
+		_, err = jwt.Parse([]byte(strToken), jwt.WithKey(jwa.ES512, raw.PublicKey), jwt.WithValidate(true))
```

Comment:
Another PR in the pipeline (#167) has to change token encryption to `ES256` for compatibility with the SciTokens library. Can you also change that here and on the token-creation side of things to prevent future merge conflicts? It'll also require that we create a new federation keypair to match, which could go in a GH issue.